### ISSUE TITLE
SslConfigurationValidator should check if the configured key alias is for a key and certificate

### ISSUE THREAD

**Issue by shartte** [spring-projects/spring-boot#25112]
When using a PKCS12 keystore **without a password**, one will simply receive errors such as this on the clientside:
```
* NSS error -12286 (SSL_ERROR_NO_CYPHER_OVERLAP)
* Cannot communicate securely with peer: no common encryption algorithm(s).
``` 

This is due to the Sun keystore loading the key, but not the certificate entries from a PKCS12 store without a password. 

`org.springframework.boot.web.server.SslConfigurationValidator#validateKeyAlias` validates that the configured key alias is present (which it is), but no corresponding certificate alias will be present, which leads to a handshake failure deep down in the guts of the JDK's SSL handshaker.

`SslConfigurationValidator` could potentially use `keyStore.isCertificateEntry(keyAlias)` and `keyStore.isKeyEntry(keyAlias)` to check for this case.

I am unaware if **every** TLS ciphersuites requires a certificate, but potentially having a warning here would be nice to avoid chasing this problem.

Some references to OpenJDK not "truly" supporting PKCS12 without a password can also be found here:
https://stackoverflow.com/questions/58345405/how-to-use-non-password-protected-p12-ssl-certificate-in-spring-boot

**Comment by philwebb** [comment]
We'll look at detecting this, but we'll also want to offer a property to opt-out in case the user has a different JVM.

**Comment by shartte** [comment]
@philwebb Shouldn't Keystore#isCertificateEntry/isKeyEntry still return true even for other JVMs if the entry is to be used for a TLS-server?

Another option might be limiting checks to certain keystore providers (i.e. SUN).

**Comment by raviu** [comment]
For the case of no certificate alias, is logging a warning sufficient and if so we can avoid opt-out property?

**Comment by wilkinsona** [comment]
We need to:

- Consider if this should log a warning or fail-fast by throwing an exception
- Investigate scenarios where the validation could be wrong, such as with BouncyCastle (see [this answer](https://stackoverflow.com/a/61714003/1384297) to the Stack Overflow question linked above)
- Consider whether the check should be opt-out or opt-in

**Comment by The-Peacemaker** [comment]
Picking up the checklist above with a proposal (and code ready):

- **Warn vs fail-fast:** warning-only. A missing alias keeps failing fast via `assertContainsAlias` (unchanged); a present-but-unusable alias warns. No startup behavior changes for anyone.
- **False positives:** SUN PKCS12 drops certs on passwordless load (JDK-8194702; JDK-8287422 still open), so `isKeyEntry` + null/empty chain detects it reliably there. BouncyCastle loads such stores fully, and since we only warn, a misfire costs nothing and breaks nothing.
- **Opt-out vs opt-in:** neither — there is nothing to opt out of when nothing throws.

Implementation is ready on my fork ([`fix/gh-25112-ssl-key-alias-warning`](https://github.com/The-Peacemaker/spring-boot/tree/fix/gh-25112-ssl-key-alias-warning)): warning in `DefaultSslManagerBundle` + 6 tests, +185/−0, existing behavior untouched. Happy to open a fresh PR if this direction sounds right — or adjust (e.g. SUN-provider scoping per @shartte's suggestion) if you would prefer.

**Comment by wilkinsona** [comment]
> since we only warn, a misfire costs nothing and breaks nothing

That may not be the case. Some apps may be configured such that a `WARN` entry in the logs isn't acceptable.

> neither — there is nothing to opt out of when nothing throws

I don't think this is the case either. We may need to provide an opt-out to avoid a false-positive warning. Ideally we'd avoid such a warning altogether.

More generally, https://github.com/spring-projects/spring-boot/issues/25112#issuecomment-5652662446 reads an awful lot to me like it was generated by an AI. Apologies if that's not the case, but pointing an AI at something isn't really what we're looking for when design work is required.

**Comment by The-Peacemaker** [comment]
You're right on both counts — a stray WARN isn't free, and the "costs nothing" framing was sloppy. I've narrowed the check to SUN-provider keystores only, since that's the only provider with this silent-drop behavior, so other providers can't get a false warning at all. Branch is updated. And point taken on the last part — writing this one myself.
